File: algo_expert/recursion/blackjack_probability.py




#  TC O(t-s) SC O(t-s)
def blackjackProbability(target, startingHand):
    memo = {}
    return rec_helper(target, startingHand, memo)


def rec_helper(target, startingHand, memo):
    if  (target - 4) <= startingHand <= target:
        return 0
    if startingHand in memo:
        return memo[startingHand]
    
    prob = 0.0 
    for card_nb in range(1, 11):
        new_startingHand = startingHand + card_nb
        if new_startingHand < target - 4:
            prob += 0.1 * rec_helper(target, new_startingHand, memo)
        elif new_startingHand > target:
            prob += 0.1
    memo[startingHand] = round(prob, 3)
    return memo[startingHand]



target = 21
startingHand = 15
res = blackjackProbability(target, startingHand)
print(f"taget={target} startingHand={startingHand}")
print(f"blackjackProbability {res}")


target = 100
startingHand = 90
res = blackjackProbability(target, startingHand)
print(f"taget={target} startingHand={startingHand}")
print(f"blackjackProbability {res}")

File: algo_expert/recursion/blackjack_probability.js


// TC O(t-s) and SC O(t-s)
function blackjackProbability(target, startingHand) {
    let memo = {}
    return recHelper(target, startingHand, memo);
}


function recHelper(target, startingHand, memo){

    if (target-4 <= startingHand && startingHand <= target){
        return 0;
    } else if (startingHand in memo){
        return memo[startingHand];
    } 

    let totalProb = 0;
    let newStartingHand;
    for (let cartNb=1; cartNb<11; cartNb++){
        newStartingHand = cartNb + startingHand
        if (newStartingHand < target-4){
            totalProb += 0.1 * recHelper(target, newStartingHand, memo);
        } else if (newStartingHand > target){
            totalProb += 0.1
        }
    }
    memo[startingHand] = Number(totalProb.toFixed(3))
    return memo[startingHand];
}

  


let target = 21;
let startingHand = 15;
let res = blackjackProbability(target, startingHand);
console.log(`target=${target} startingHand=${startingHand}`);
console.log(`blackjackProbability ${res}`);

target = 100;
startingHand = 90;
res = blackjackProbability(target, startingHand);
console.log(`target=${target} startingHand=${startingHand}`);
console.log(`blackjackProbability ${res}`);